### PR TITLE
Implement memlog store operations

### PR DESCRIPTION
**Pull request by urso**
- Enhancement

## What does this PR do?

Implement store operation that get/set key value pairs from
the in-memory store, and log update operations.

This change also include unit tests and store compliance tests.

The addition of the statestore package is split up into multiple
changeset to ease review. The final version of the package can be found
[here](https://github.com/urso/beats/tree/fb-input-v2-combined/libbeat/statestore).

Once finalized, the libbeat/statestore package contains:
- The statestore frontend and interface for use within Beats
- Interfaces for the store backend
- A common set of tests store backends need to support
- a storetest package for testing new features that require a store. The
  testing helpers use map[string]interface{} that can be initialized or
  queried after the test run for validation purposes.
- The default memlog backend + tests

## Why is it important?

The statestore introduces a simple key-value store to Beats. The statestore will be used to replace the registry in filebeat in the future.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
~~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## Author's Checklist

- [ ]

## How to test this PR locally

`go test`

## Related issues

- Relates elastic/beats#15324

### CODE REVIEW

File: libbeat/statestore/backend/memlog/memlog.go [hunk 1]
```diff
@@ -54,6 +54,9 @@ type Settings struct {
 	// Checkpoint predicate that can trigger a registry file rotation.  If not
 	// configured, memlog will automatically trigger a checkpoint every 10MB.
 	Checkpoint CheckpointPredicate
+
+	// If set memlog will not check the version of the meta file.
+	IgnoreVersionCheck bool
```
Comment (urso) [inline libbeat/statestore/backend/memlog/memlog.go:59]: This is currently required by the filebeat migration (PR not open yet). When migrating the old store we create a temporary store in the same directory and convert all entries before updating the meta-data file (migration supports 3 different formats, which made the process a little hairy).

File: go.mod [hunk 2]
```diff
@@ -163,7 +163,7 @@ require (
 	golang.org/x/sys v0.0.0-20200202164722-d101bd2416d5
 	golang.org/x/text v0.3.2
 	golang.org/x/time v0.0.0-20191024005414-555d28b269f0
-	golang.org/x/tools v0.0.0-20200630154851-b2d8b0336632
+	golang.org/x/tools v0.0.0-20200630223951-c138986dd9b9
```
Comment (urso) [inline go.mod:166]: yay, go modules and verion pinning :/

File: libbeat/statestore/backend/memlog/memlog_test.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+func TestLoadVersion1(t *testing.T) {
+	dataHome := "testdata/1"
+
+	list, err := ioutil.ReadDir(dataHome)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cases := list[:0]
```
Comment (faec) [inline libbeat/statestore/backend/memlog/memlog_test.go:67]: I'm not sure I understand this construction -- this should be equivalent to `cases := []os.FileInfo{}`, right? Is this a type inference shortcut in the spirit of c++ `auto`, or something else?

File: libbeat/statestore/backend/memlog/store.go [hunk 4]
```diff
@@ -56,51 +59,201 @@ type entry struct {
...
+		if err != nil {
+			return nil, err
+		}
+
+		err = writeMetaFile(home, mode)
+		if err != nil {
+			return nil, err
+		}
+	} else if !fi.Mode().IsDir() {
+		return nil, fmt.Errorf("'%v' is no directory", home)
```
Comment (faec) [inline libbeat/statestore/backend/memlog/store.go:75]: grammar nit: "no" -> "not a"

File: libbeat/statestore/backend/memlog/store.go [hunk 5]
```diff
@@ -56,51 +59,201 @@ type entry struct {
...
+	var entries uint
+	memstore := memstore{tbl}
+	txid, entries, err = loadLogFile(&memstore, txid, home)
+	logp.Info("Finished loading transaction log file for '%v'. Active transaction id=%v", home, txid)
+
+	if err != nil {
+		// Error indicates the log file was incomplete or corrupted.
+		// Anyways, we already have the table in a valid state and will
+		// continue opening the store from here.
+		logp.Warn("Incomplete or corrupted log file in %v. Continue with last known complete and consistend state. Reason: %v", home, err)
```
Comment (faec) [inline libbeat/statestore/backend/memlog/store.go:130]: typo: "consistent"

File: libbeat/statestore/backend/memlog/memlog.go [hunk 6]
```diff
@@ -110,7 +113,7 @@ func (r *Registry) Access(name string) (backend.Store, error) {
 	home := filepath.Join(r.settings.Root, name)
 	fileMode := r.settings.FileMode
 	bufSz := r.settings.BufferSize
-	store, err := openStore(logger, home, fileMode, bufSz, r.settings.Checkpoint)
+	store, err := openStore(logger, home, fileMode, bufSz, !r.settings.IgnoreVersionCheck, r.settings.Checkpoint)
```
[the remark this answers is not in the document]
Comment (urso) [reply libbeat/statestore/backend/memlog/memlog.go:116]: Bad habits :)
Update to ignoreVersionCheck.

File: libbeat/statestore/backend/memlog/memlog_test.go (hunk 3 above)
Comment (urso) [reply libbeat/statestore/backend/memlog/memlog_test.go:67]: >  this should be equivalent to cases := []os.FileInfo{}, right?

kinda. As we have slices the construct creates a slice of length 0, but with the old capacity.
It's an in-place filtering loop for removing files from the list (e.g. the Readme.md). See: https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating

But yeah, this is me being lazy :)

**Comment by urso** [comment]
Filebeat tests are all green. CI failed notifying github

**Pull request merged**